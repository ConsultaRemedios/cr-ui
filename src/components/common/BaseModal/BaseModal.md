A raw Modal component. This component does not control the style inside the modal. It just
renders a Modal with a close button (an icon).

### Usage normal
```js
new Vue({
  data(){
    return { showModal: false };
  },
  methods: {
    openModal() {
      this.showModal = true;
    },

    closeModal() {
      this.showModal = false;
    },
  },
  template: `
    <div>
      <BaseModal v-if="showModal" @close="closeModal">
        <div :style="{ padding: '20px' }">
          <h2>Base Modal</h2>
          <p>This is a raw base modal. It's uggly, I know.</p>
        </div>
      </BaseModal>

      <BaseButton @click="openModal">Open Modal</BaseButton>
    </div>
  `
})
```
### Usage footer modal
```js
new Vue({
  data(){
    return { showModal: false };
  },
  methods: {
    openModal() {
      this.showModal = true;
    },

    closeModal() {
      this.showModal = false;
    },
  },
  template: `
    <div>
      <BaseModal v-if="showModal" @close="closeModal" type="footer">
        <div :style="{ padding: '20px' }">
          <h2>Base Modal</h2>
          <p>This is a raw base modal. It's uggly, I know.</p>
        </div>
      </BaseModal>

      <BaseButton @click="openModal">Open Modal</BaseButton>
    </div>
  `
})
```

### Dismissible
You can set dismissible prop as `true` if you don't want a close button.
Also, The BaseModal component will not trigger `close` event when clicking on overlay.

```js
new Vue({
  data(){
    return { showModal: false };
  },
  methods: {
    openModal() {
      this.showModal = true;
    },

    closeModal() {
      this.showModal = false;
    },
  },
  template: `
    <div>
      <BaseModal v-if="showModal" :dismissible="false">
        <div :style="{ padding: '20px' }">
          <h2>Base Modal</h2>
          <p>This is a raw base modal. It's uggly, I know.</p>

          <BaseButton @click="closeModal">Close Modal</BaseButton>
        </div>
      </BaseModal>

      <BaseButton @click="openModal">Open Modal</BaseButton>
    </div>
  `
})
```
