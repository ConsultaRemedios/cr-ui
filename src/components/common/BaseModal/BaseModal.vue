<template>
  <transition
    :enter-class="$style.enterTransition"
    :leave-active-class="$style.leaveActiveTransition"
  >
    <div :class="[$style.overlay, 'cr-base-modal__overlay', customCssClass]" @click="onClickOverlay" data-modal-overlay>
      <div :class="[$style.container, 'cr-base-modal__container']" data-modal-container>
        <div :class="[$style.content, 'cr-base-modal__content']" data-modal-content>
          <button v-if="dismissible" data-modal-close
            :class="[$style.closeButton, 'cr-base-modall__close']"
            @click="$emit('close')"
          >
            <span :class="$style.closeButtonLabel">Fechar</span>
          </button>

          <slot></slot>
        </div>
      </div>
    </div>
  </transition>
</template>

<script>
  import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';

  export default {
    name: 'BaseModal',
    props: {
      dismissible: {
        type: Boolean,
        default: true,
      },

      customCssClass: {
        type: Array,
        default: () => [],
      },
    },

    methods: {
      onClickOverlay(e) {
        const isOverlayTheTarget = e.target.querySelector('[data-modal-content]');

        if (isOverlayTheTarget && this.dismissible) {
          this.$emit('close');
        }
      },
    },

    mounted() {
      disableBodyScroll(this.$el);
    },

    destroyed() {
      enableBodyScroll(this.$el);
    },
  };
</script>

<style module>
  .overlay {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background: rgba(0, 0, 0, 0.6);
    z-index: 9999;
    overflow: auto;

    transition: opacity 0.2s ease-out;
    opacity: 1;
  }

  .container {
    min-width: 100%;
    min-height: 100%;
    display: flex;
    align-items: center;
    justify-content: center;
    padding: 20px;
  }

  .content {
    background: #FFF;
    position: relative;
    border-radius: 4px;
    transition: transform 0.2s ease-out;
    transform: scale(1);
  }

  .closeButton {
    position: absolute;
    top: 15px;
    right: 15px;
    cursor: pointer;
    padding: 5px;
    background: none;
    border: 0;
  }

  .closeButton::after {
    content: '';
    display: block;
    width: 14px;
    height: 14px;
    background: url('./../../../assets/close-icon.svg');
  }

  .closeButtonLabel {
    display: block;
    font-size: 0;
    height: 1px;
    overflow: hidden;
  }

  .enterTransition,
  .leaveActiveTransition {
    opacity: 0;
  }

  .enterTransition .content,
  .leaveActiveTransition .content {
    transform: scale(1.05);
  }

  @media (min-width: 768px) {
    .closeButton {
      top: 20px;
      right: 20px;
    }
  }
</style>
