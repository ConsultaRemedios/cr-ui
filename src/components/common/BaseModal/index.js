import BaseModal from './BaseModal.vue';

export default BaseModal;
