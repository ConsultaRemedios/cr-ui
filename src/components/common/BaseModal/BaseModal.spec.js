import { mount } from '@vue/test-utils';
import { disableBodyScroll, enableBodyScroll } from 'body-scroll-lock';
import snapshotDiff from 'snapshot-diff';
import BaseModal from './../BaseModal';

jest.mock('body-scroll-lock', () => ({
  disableBodyScroll: jest.fn(),
  enableBodyScroll: jest.fn()
}));

const mountModal = (opts) => {
  disableBodyScroll.mockClear();
  enableBodyScroll.mockClear();

  const options = opts || {};
  const defaultOptions = {
    slots: {
      default: ['Modal content here'],
    },
  };
  return mount(BaseModal, { ...defaultOptions, ...options });
}

describe('BaseModal Component', () => {
  describe('Default props value', () => {
    describe('#dismissible', () => {
      it('returns true by default', () => {
        const wrapper = mount(BaseModal);
        expect(wrapper.vm.dismissible).toBe(true);
      });
    });
  });

  describe('render', () => {
    it('does not render close button when dismissible is equal to "false"', () => {
      const wrapper = mountModal({
        propsData: { dismissible: false }
      });

      expect(wrapper.find('button[data-modal-close]').exists()).toBe(false);
    });

    it('does render close button when dismissible is equal to "true"', () => {
      const wrapper = mountModal({
        propsData: { dismissible: true }
      });

      expect(wrapper.find('button[data-modal-close]').exists()).toBe(true);
    });
  });

  describe('When mounted', () => {
    it('calls disableBodyScroll', () => {
      const wrapper = mountModal();
      expect(disableBodyScroll).toHaveBeenCalledTimes(1);
      expect(disableBodyScroll).toHaveBeenCalledWith(wrapper.vm.$el);
    })
  });

  describe('When destroyed', () => {
    it('calls disableBodyScroll', () => {
      const wrapper = mountModal();
      expect(enableBodyScroll).not.toHaveBeenCalled();

      wrapper.destroy();
      expect(enableBodyScroll).toHaveBeenCalledTimes(1);
      expect(enableBodyScroll).toHaveBeenCalledWith(wrapper.vm.$el);
    })
  });

  describe('When click on close button', () => {
    it('emits close event', () => {
      const wrapper = mountModal();
      jest.spyOn(wrapper.vm, '$emit');

      wrapper.find('button[data-modal-close]').trigger('click');
      expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
    });
  });

  describe('When click on modal element', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mountModal();
      jest.spyOn(wrapper.vm, '$emit');
    });

    afterEach(() => {
      wrapper.vm.$emit.mockRestore();
    });

    describe('When the target is the overlay', () => {
      describe('and dismissible prop is "true"', () => {
        it('emits close event', () => {
          wrapper.find('[data-modal-overlay]').trigger('click');
          expect(wrapper.vm.$emit).toHaveBeenCalledTimes(1);
        });
      });

      describe('and dismissible prop is "false"', () => {
        it('does not emit close event', () => {
          wrapper = mountModal({
            propsData: { dismissible: false }
          });

          jest.spyOn(wrapper.vm, '$emit');

          wrapper.find('[data-modal-overlay]').trigger('click');
          expect(wrapper.vm.$emit).not.toHaveBeenCalled();
        });
      });
    });

    describe('When the target is the modal content', () => {
      it('does not emit close event', () => {
        wrapper.find('[data-modal-content]').trigger('click');
        expect(wrapper.vm.$emit).not.toHaveBeenCalled();
      });
    });
  });

  describe('When rendered', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(BaseModal, {
        slots: {
          default: ['<div>Modal content here</div>'],
        },
      });
    });

    it('matches to snapshot', () => {
      expect(wrapper.vm.$el).toMatchSnapshot();
    });

    it('matches to snapshot when props "customCssClass" is passed', () => {
      const localWrapper = mount(BaseModal, {
        slots: {
          default: ['<div>Modal content here</div>'],
        },
        propsData: { customCssClass: ['some-weird-css-class'] }
      });

      expect(snapshotDiff(wrapper.element, localWrapper.element)).toMatchSnapshot();
    });
  });
});
