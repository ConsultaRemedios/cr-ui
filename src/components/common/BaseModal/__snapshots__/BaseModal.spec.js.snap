// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseModal Component When rendered matches to snapshot 1`] = `
<div
  class="overlay cr-base-modal__overlay"
  data-modal-overlay=""
  enter-class="enterTransition"
  leave-active-class="leaveActiveTransition"
>
  <div
    class="container cr-base-modal__container"
    data-modal-container=""
  >
    <div
      class="content cr-base-modal__content"
      data-modal-content=""
    >
      <button
        class="closeButton cr-base-modall__close"
        data-modal-close=""
      >
        <span
          class="closeButtonLabel"
        >
          Fechar
        </span>
      </button>
       
      <div>
        Modal content here
      </div>
    </div>
  </div>
</div>
`;

exports[`BaseModal Component When rendered matches to snapshot when props "customCssClass" is passed 1`] = `
Snapshot Diff:
- First value
+ Second value

@@ -1,7 +1,7 @@
  <div
-   class="overlay cr-base-modal__overlay"
+   class="overlay cr-base-modal__overlay some-weird-css-class"
    data-modal-overlay=""
    enter-class="enterTransition"
    leave-active-class="leaveActiveTransition"
  >
    <div
`;
